tree/ntuple/v7/inc/ROOT/RNTupleMerger.hxx

```diff
@@ -34,10 +46,67 @@ class RFieldMerger {
 private:
    /// The merged field descriptor
    RFieldDescriptor fMergedField = RFieldDescriptor();
+
 public:
    static RResult<RFieldMerger> Merge(const RFieldDescriptor &lhs, const RFieldDescriptor &rhs);
 };
 
+// clang-format off
+/**
+ * \class ROOT::Experimental::RNTupleMerger
+ * \ingroup NTuple
+ * \brief Given a set of RPageSources merge them into an RPageSink
+ */
+// clang-format on
+class RNTupleMerger {
+
+public:
+   /// Merge a given set of sources into the destination
+   void Merge(std::vector<std::unique_ptr<RPageSource>> &sources, std::unique_ptr<RPageSink> &destination);
+
+private:
+   // Struct to hold column and field descriptors
+   struct RColumnInfo {
+      const RColumnDescriptor &fColumnDesc;
+      const RFieldDescriptor &fFieldDesc;
+      const std::uint64_t fIndex;
+
+      RColumnInfo(const RColumnDescriptor &columnDesc, const RFieldDescriptor &fieldDesc, const std::uint64_t &index)
+         : fColumnDesc(columnDesc), fFieldDesc(fieldDesc), fIndex(index)
+      {
+      }
+   };
+
+   /// Recursively add columns from a given filed
+   void
+   AddColumnsFromField(std::vector<RColumnInfo> &vec, const RNTupleDescriptor &desc, const RFieldDescriptor &fieldDesc)
+   {
+      for (const auto &field : desc.GetFieldIterable(fieldDesc)) {
+         for (const auto &column : desc.GetColumnIterable(field)) {
+            const std::string name = field.GetFieldName() + "." + std::to_string(column.GetIndex());
+            if (!m_indexMap.count(name)) { // contains as of C++20
+               m_indexMap[name] = m_indexMap.size();
+            }
+            vec.emplace_back(column, field, m_indexMap.at(name));
```

Comment:
I think this assumes that iteration through a source discovers the columns in the same order than they are created in the sink, and that may not be true. I think it would be safer to recursively:
- Iterate over all direct sub fields of the current parent field
- Find that subfield by name and parent ID in the destination (`RNTupleDescriptor::FindFieldId`)
  - Ensure that the fields have the same type and version 
- Loop over the source/destination columns of the fields; ensure that the columns are identical
- Map source column ID to destination column ID
- At the end: count number of source fields and target fields (to make sure that the target does not have fields/columns not present in the source)

Reply:
The current implementation indeed needs to be improved. Right now, when collect the columns from the first source, we build an internal map that holds the name (as described above) and an index (basically a counter). Then, when we loop over the remaining sources, we look up the index by name and use that. This way even if the input column ids are different between the sources, things work OK (actually the test checks this, where I change the order `foo` and `bar` are inserted into the merged sources). This, however, presumes all sources have the same columns etc. I'll go over your suggestion and improve the implementation. Having said that, how should the logic behave if it encounters incompatible sources? Should we just bail?

Reply:
I think that's a decent first approach: if the sources are incompatible, we through an exception.

Reply:
After this mornings chat I reworked the logic here a bit. Now, we do a more stricter check in terms of the column contents between the sources. We basically take the first source as the baseline and expect the remaining sources to have the same exact content (order can be different, of course). If not, we raise an exception.

Philippe's comments about taking the intersection of the columns across the sources and allowing the user to merge sources with the option of dropping certain columns are still relevant and not addressed in this current implementation. Can that be addressed in the future Jacob, or would you rather do that in this PR? I was not super sure how that selection would be propagated through the chain, though (for example, do we have something analogous to `TTree::SetBranchStatus` in `RNTuple`, which can possibly be used to expose only "active" columns when we collect them for merging).